# Chapter 7 - Development Method

> The development technique, also known as the chain ladder technique, is one of
> the most frequently used methodologies for estimating unpaid claims.
>
> -- Friedland, p84

This chapter covers the foundational development method (or chainladder) method. In the chainladder package, this is implemented in the ``Development`` estimator.  The  develops ultimate claims for each accident year based on the assumption that future claims will develop similarly to prior years' claims. The development method consists of the following steps:  

1. Compile claims data in a development triangle;
1. Calculate age-to-age factors;
1. Calculate averages of the age-to-age factors;
1. Select claim development factors (CDFs);
1. Select the tail factor;
1. Calculate cumulative CDFs;

In this chapter we will recreate the following Friedland Chapter 7 exhibits using the chainladder package:

- **Exhibit I** — U.S. Industry Auto
- **Exhibit II** — XYZ Insurer (Auto BI)
- **Exhibit III** — U.S. PP Auto (impact of changing conditions)
- **Exhibit IV** — U.S. Auto (impact of change in product mix)

In [12]:
import chainladder as cl
from IPython.display import display, HTML, Markdown
import numpy as np
import pandas as pd


pd.set_option("display.max_columns", None)
pd.set_option("display.width", 1000)

import warnings

warnings.filterwarnings(
    "ignore",
    message=r"Some exclusions have been ignored\..*link ratio\(s\) is required.*",
    category=UserWarning,
)


#pd.options.display.float_format = '{:,.0f}'.format

## Exhibit I — U.S. Industry Auto (Impact of Changing Conditions)

In Exhibit I,  Friedland demonstrates seven steps of the development method by examining industry-aggregated accident year claim development data for U.S. private passenger auto insurance. 

### Exhibit I, Sheet 1 - U.S. Industry Auto - Reported Claims ($000)

We begin by generating Sheet 1, Parts 1-4 which contain the data triangle and age-to-age factors for US Industry Auto reported claims under the increasing claim ratio scenario (p. 106).

#### PART 1 - Data Triangle

We have already imported the necessary packages loading the ``Triangle`` at the top of p106. Let's take a look at the ``Triangle`` we just loaded.


In [3]:
tri = cl.load_sample('friedland_us_industry_auto')
tri['Reported Claims']

,12,24,36,48,60,72,84,96,108,120
1998,"37,017,487","43,169,009","45,568,919","46,784,558","47,337,318","47,533,264","47,634,419","47,689,655","47,724,678","47,742,304"
1999,"38,954,484","46,045,718","48,882,924","50,219,672","50,729,292","50,926,779","51,069,285","51,163,540","51,185,767",
2000,"41,155,776","49,371,478","52,358,476","53,780,322","54,303,086","54,582,950","54,742,188","54,837,929",,
2001,"42,394,069","50,584,112","53,704,296","55,150,118","55,895,583","56,156,727","56,299,562",,,
2002,"44,755,243","52,971,643","56,102,312","57,703,851","58,363,564","58,592,712",,,,
2003,"45,163,102","52,497,731","55,468,551","57,015,411","57,565,344",,,,,
2004,"45,417,309","52,640,322","55,553,673","56,976,657",,,,,,
2005,"46,360,869","53,790,061","56,786,410",,,,,,,
2006,"46,582,684","54,641,339",,,,,,,,
2007,"48,853,563",,,,,,,,,


#### PART 2 - Age-to-Age Factors

To calculate age-to-age factors, use the ``age-to-age`` attribute of the ``Triangle``.

In [4]:
tri['Reported Claims'].age_to_age.round(decimals = 3)

,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
1998,1.1660,1.0560,1.0270,1.0120,1.0040,1.0020,1.0010,1.0010,1.0000
1999,1.1820,1.0620,1.0270,1.0100,1.0040,1.0030,1.0020,1.0000,
2000,1.2000,1.0610,1.0270,1.0100,1.0050,1.0030,1.0020,,
2001,1.1930,1.0620,1.0270,1.0140,1.0050,1.0030,,,
2002,1.1840,1.0590,1.0290,1.0110,1.0040,,,,
2003,1.1620,1.0570,1.0280,1.0100,,,,,
2004,1.1590,1.0550,1.0260,,,,,,
2005,1.1600,1.0560,,,,,,,
2006,1.1730,,,,,,,,


#### PART 3 - Average Age-to-Age Factors

To calculate the average age-to-age factors, we will use the ``Development`` estimator to ``fit_transform`` the original ``Triangle``. This calculates the averages but also preserves the ability to apply other estimators later. The specific choices of average paramters (n_period, etc.) are provided to ``Development``. The attribute for the calculated average age-to-age factors is the ``ldf_``.

In [5]:
display(Markdown("##### Reported Claims - Simple Average (Latest 5)"))
reported_simple_5 = cl.Development(n_periods=5, average='simple').fit_transform(tri['Reported Claims'])
display(reported_simple_5.ldf_.round(3))

display(Markdown("##### Reported Claims - Simple Average (Latest 3)"))
reported_simple_3 = cl.Development(n_periods=3, average='simple').fit_transform(tri['Reported Claims'])
display(reported_simple_3.ldf_.round(3))

display(Markdown("##### Reported Claims - Medial Average (Latest 5 x 1)"))
reported_medial_5x1 = cl.Development(n_periods=5, average='simple', drop_high = 1, drop_low = 1).fit_transform(tri['Reported Claims'])
display(reported_medial_5x1.ldf_.round(decimals = 3))

display(Markdown("##### Reported Claims - Volume-weighted Average (Latest 5)"))
reported_volume_5 = cl.Development(n_periods=5, average='volume').fit_transform(tri['Reported Claims'])
display(reported_volume_5.ldf_.round(decimals = 3))

display(Markdown("##### Reported Claims - Volume-weighted Average (Latest 3)"))
reported_volume_3 = cl.Development(n_periods=3, average='volume').fit_transform(tri['Reported Claims'])
display(reported_volume_3.ldf_.round(decimals = 3))

display(Markdown("##### Reported Claims - Geometric Average (Latest 4)"))
reported_volume_3 = cl.Development(n_periods=4, average='geometric').fit_transform(tri['Reported Claims'])
display(reported_volume_3.ldf_.round(decimals = 3))



##### Reported Claims - Simple Average (Latest 5)

,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
(All),1.1680,1.0580,1.0270,1.0110,1.0040,1.0030,1.0020,1.0010,1.0000


##### Reported Claims - Simple Average (Latest 3)

,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
(All),1.1640,1.0560,1.0270,1.0120,1.0050,1.0030,1.0020,1.0010,1.0000


##### Reported Claims - Medial Average (Latest 5 x 1)

,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
(All),1.1650,1.0570,1.0270,1.0100,1.0040,1.0030,1.0020,1.0010,1.0000


##### Reported Claims - Volume-weighted Average (Latest 5)

,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
(All),1.1680,1.0580,1.0270,1.0110,1.0040,1.0030,1.0020,1.0010,1.0000


##### Reported Claims - Volume-weighted Average (Latest 3)

,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
(All),1.1640,1.0560,1.0270,1.0120,1.0050,1.0030,1.0020,1.0010,1.0000


##### Reported Claims - Geometric Average (Latest 4)

,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
(All),1.1640,1.0570,1.0270,1.0110,1.0040,1.0030,1.0020,1.0010,1.0000


#### PART 4 - Selected Age-to-Age Factors

For the prior selected, we need to create a hard-coded pattern, using the ``DevelopmentConstant`` estimator. In a production workflow, you can save the development pattern from the prior analysis and load for reference in a subsequent analysis. 

We will also be using the ``TailConstant`` estimator to add a tail factor to selected development patterns. We add the tail factor to a transformed ``Triangle`` (i.e. applying ``fit_transform`` of the ``Development`` estimator to a ``Triangle``) by using ``fit_transform`` once again.

In [6]:
reported_prior_method =  cl.DevelopmentConstant(
    patterns = {
        12:1.16, 
        24:1.057, 
        36:1.028, 
        48:1.012, 
        60:1.005, 
        72:1.003, 
        84:1.001, 
        96:1.001, 
        108:1.000
        }, 
        style='ldf'
        )

reported_prior_ft = reported_prior_method.fit_transform(tri['Reported Claims'])
reported_tail_method = cl.TailConstant(
    tail = 1,
    projection_period = 0
    )
reported_prior_selected = reported_tail_method.fit_transform(reported_prior_ft)
display(Markdown("##### Prior Selected"))
display(reported_prior_selected.ldf_)

##### Prior Selected

,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120,120-132
(All),1.1600,1.0570,1.0280,1.0120,1.0050,1.0030,1.0010,1.0010,1.0000,1.0000


Next, we can select some factors. We can also reuse the `TailConstant` from the previous step. This is fairly common in practice, as tail factors are selected less frequently than the development pattern itself, so need to be carried from analysis to analysis.

In [7]:
reported_selected_pattern = reported_tail_method.fit_transform(reported_simple_3)
display(Markdown("##### Selected"))
display(reported_selected_pattern.ldf_)

##### Selected

,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120,120-132
(All),1.1641,1.0559,1.0273,1.0115,1.0046,1.0028,1.0016,1.0006,1.0004,1.0000


The Development estimator has a ``cdf_`` attribute that will automatically multiply age-to-age factors cumulatively into age-to-ultimate factors. The Friedland text uses the rounded LDF to calculate CDF. This can be achieved in this package by using the ``incr_to_cum()`` method of the rounded age-to-age factors.

In [8]:
from contextlib import contextmanager
from chainladder.core.display import TriangleDisplay

@contextmanager
def unrounded_triangle_display():
    original = TriangleDisplay._get_format_str
    TriangleDisplay._get_format_str = staticmethod(lambda data: "{}")
    try:
        yield
    finally:
        TriangleDisplay._get_format_str = staticmethod(original)

display(Markdown("##### CDF to Ultimate (without rounding)"))
with unrounded_triangle_display():
    display(reported_selected_pattern.cdf_)

display(Markdown("##### CDF to Ultimate (with rounding)"))
reported_selected_cdf = reported_selected_pattern.ldf_.round(decimals = 3).incr_to_cum().round(decimals = 3)
display(reported_selected_cdf)


##### CDF to Ultimate (without rounding)

,12-Ult,24-Ult,36-Ult,48-Ult,60-Ult,72-Ult,84-Ult,96-Ult,108-Ult,120-Ult
(All),1.2899769284284377,1.1081386197639869,1.0494934575145254,1.0215544575100222,1.0099084980723658,1.0053002172823449,1.0025401850030309,1.0009539548490918,1.000369326745379,1.0


##### CDF to Ultimate (with rounding)

,12-Ult,24-Ult,36-Ult,48-Ult,60-Ult,72-Ult,84-Ult,96-Ult,108-Ult,120-Ult
(All),1.2920,1.1100,1.0510,1.0230,1.0110,1.0060,1.0030,1.0010,1.0000,1.0000


To calculate % reported, we will use ``Triangle`` manipulation from Chapter 5 directly on the development pattern (which is also a ``Triangle``).

In [9]:
display(Markdown("##### Percent Reported"))
display((1 / reported_selected_cdf).round(decimals = 3))

##### Percent Reported

,12-Ult,24-Ult,36-Ult,48-Ult,60-Ult,72-Ult,84-Ult,96-Ult,108-Ult,120-Ult
(All),0.7740,0.9010,0.9510,0.9780,0.9890,0.9940,0.9970,0.9990,1.0000,1.0000


### Exhibit I, Sheet 2 (p. 107)

Moving onto the next page, all the calculations are identical to the previous page. We will manually repeat the same code. In a production workflow, commonly repeated methods and selections can be streamlined, which we will demonstrate in Exhibit II.

In [10]:
display(Markdown("#### Part 1 - Data Triangle"))
display(tri['Paid Claims'])

display(Markdown("#### Part 2 - Age-to-Age Factors"))
display(tri['Paid Claims'].age_to_age.round(decimals = 3))

display(Markdown("#### Part 3 - Average Age-to-Age Factors"))

display(Markdown("##### Simple Average"))
display(Markdown("##### Latest 5"))
paid_simple_5 = cl.Development(n_periods=5, average='simple').fit_transform(tri['Paid Claims'])
display(paid_simple_5.ldf_.round(decimals = 3))

display(Markdown("##### Latest 3"))
paid_simple_3 = cl.Development(n_periods=3, average='simple').fit_transform(tri['Paid Claims'])
display(paid_simple_3.ldf_.round(decimals = 3))

display(Markdown("##### Medial Average"))
display(Markdown("##### Latest 5 x 1"))
paid_medial_5x1 = cl.Development(n_periods=5, average='simple',drop_high = 1, drop_low = 1, preserve=1).fit_transform(tri['Paid Claims'])
display(paid_medial_5x1.ldf_)

display(Markdown("##### Volume-weighted Average"))
display(Markdown("##### Latest 5"))
paid_volume_5 = cl.Development(n_periods=5, average='volume').fit_transform(tri['Paid Claims'])
display(paid_volume_5.ldf_.round(decimals = 3) )

display(Markdown("##### Latest 3"))
paid_volume_3 = cl.Development(n_periods=3, average='volume').fit_transform(tri['Paid Claims'])
display(paid_volume_3.ldf_.round(decimals = 3) )

display(Markdown("##### Geometric Average"))
display(Markdown("##### Latest 4"))
paid_geometric_4 = cl.Development(n_periods=4, average='geometric').fit_transform(tri['Paid Claims'])
display(paid_geometric_4.ldf_.round(decimals = 3))

display(Markdown("#### Part 4 - Selected Age-to-Age Factors"))
display(Markdown("#### Prior Selected"))

paid_prior_method =  cl.DevelopmentConstant(
    patterns = {
    12:1.707, 
    24:1.189, 
    36:1.091, 
    48:1.044, 
    60:1.019, 
    72:1.01, 
    84:1.005, 
    96:1.003, 
    108:1.001
    }, 
    style='ldf'
    )
paid_prior_ft = paid_prior_method.fit_transform(tri['Paid Claims'])
paid_tail_method = cl.TailConstant(
    tail = 1.002,
    projection_period = 0
    )
paid_prior_selected = paid_tail_method.fit_transform(paid_prior_ft)
display(paid_prior_selected.ldf_)

display(Markdown("##### Selected"))
paid_selected_pattern = paid_tail_method.fit_transform(paid_simple_3)
display(paid_selected_pattern)

display(Markdown("##### CDF to Ultimate"))
paid_selected_cdf = paid_selected_pattern.ldf_.round(decimals = 3).incr_to_cum().round(decimals = 3)
display(paid_selected_cdf)

display(Markdown("##### Percent Reported"))
display((1 / paid_selected_cdf).round(decimals = 3))


#### Part 1 - Data Triangle

,12,24,36,48,60,72,84,96,108,120
1998,"18,539,254","33,231,039","40,062,008","43,892,039","45,896,535","46,765,422","47,221,322","47,446,877","47,555,456","47,644,187"
1999,"20,410,193","36,090,684","43,259,402","47,159,241","49,208,532","50,162,043","50,625,757","50,878,808","51,000,534",
2000,"22,120,843","38,976,014","46,389,282","50,562,385","52,735,280","53,740,101","54,284,334","54,533,225",,
2001,"22,992,259","40,096,198","47,767,835","52,093,916","54,363,436","55,378,801","55,878,421",,,
2002,"24,092,782","41,795,313","49,903,803","54,352,884","56,754,376","57,807,215",,,,
2003,"24,084,451","41,399,612","49,070,332","53,584,201","55,930,654",,,,,
2004,"24,369,770","41,489,863","49,236,678","53,774,672",,,,,,
2005,"25,100,697","42,702,229","50,644,994",,,,,,,
2006,"25,608,776","43,606,497",,,,,,,,
2007,"27,229,969",,,,,,,,,


#### Part 2 - Age-to-Age Factors

,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
1998,1.7920,1.2060,1.0960,1.0460,1.0190,1.0100,1.0050,1.0020,1.0020
1999,1.7680,1.1990,1.0900,1.0430,1.0190,1.0090,1.0050,1.0020,
2000,1.7620,1.1900,1.0900,1.0430,1.0190,1.0100,1.0050,,
2001,1.7440,1.1910,1.0910,1.0440,1.0190,1.0090,,,
2002,1.7350,1.1940,1.0890,1.0440,1.0190,,,,
2003,1.7190,1.1850,1.0920,1.0440,,,,,
2004,1.7030,1.1870,1.0920,,,,,,
2005,1.7010,1.1860,,,,,,,
2006,1.7030,,,,,,,,


#### Part 3 - Average Age-to-Age Factors

##### Simple Average

##### Latest 5

,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
(All),1.7120,1.1890,1.0910,1.0440,1.0190,1.0100,1.0050,1.0020,1.0020


##### Latest 3

,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
(All),1.7020,1.1860,1.0910,1.0440,1.0190,1.0090,1.0050,1.0020,1.0020


##### Medial Average

##### Latest 5 x 1

,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
(All),1.7081,1.1880,1.0908,1.0436,1.0189,1.0095,1.0048,1.0023,1.0019


##### Volume-weighted Average

##### Latest 5

,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
(All),1.7120,1.1890,1.0910,1.0440,1.0190,1.0100,1.0050,1.0020,1.0020


##### Latest 3

,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
(All),1.7020,1.1860,1.0910,1.0440,1.0190,1.0090,1.0050,1.0020,1.0020


##### Geometric Average

##### Latest 4

,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
(All),1.7060,1.1880,1.0910,1.0440,1.0190,1.0100,1.0050,1.0020,1.0020


#### Part 4 - Selected Age-to-Age Factors

#### Prior Selected

,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120,120-132
(All),1.7070,1.1890,1.0910,1.0440,1.0190,1.0100,1.0050,1.0030,1.0010,1.0020


##### Selected

,12,24,36,48,60,72,84,96,108,120
1998,"18,539,254","33,231,039","40,062,008","43,892,039","45,896,535","46,765,422","47,221,322","47,446,877","47,555,456","47,644,187"
1999,"20,410,193","36,090,684","43,259,402","47,159,241","49,208,532","50,162,043","50,625,757","50,878,808","51,000,534",
2000,"22,120,843","38,976,014","46,389,282","50,562,385","52,735,280","53,740,101","54,284,334","54,533,225",,
2001,"22,992,259","40,096,198","47,767,835","52,093,916","54,363,436","55,378,801","55,878,421",,,
2002,"24,092,782","41,795,313","49,903,803","54,352,884","56,754,376","57,807,215",,,,
2003,"24,084,451","41,399,612","49,070,332","53,584,201","55,930,654",,,,,
2004,"24,369,770","41,489,863","49,236,678","53,774,672",,,,,,
2005,"25,100,697","42,702,229","50,644,994",,,,,,,
2006,"25,608,776","43,606,497",,,,,,,,
2007,"27,229,969",,,,,,,,,


##### CDF to Ultimate

,12-Ult,24-Ult,36-Ult,48-Ult,60-Ult,72-Ult,84-Ult,96-Ult,108-Ult,120-Ult
(All),2.3900,1.4040,1.1840,1.0850,1.0400,1.0200,1.0110,1.0060,1.0040,1.0020


##### Percent Reported

,12-Ult,24-Ult,36-Ult,48-Ult,60-Ult,72-Ult,84-Ult,96-Ult,108-Ult,120-Ult
(All),0.4180,0.7120,0.8450,0.9220,0.9620,0.9800,0.9890,0.9940,0.9960,0.9980


### Exhibit I, Sheet 3 (p. 108)

This is a common report layout for reserving analyses. Some ``Pandas`` manipulation is needed to retrieve all the figures from the transformed ``Triangle`` objects and achieve the tabular look. We will create a function to reuse the manipulation throughout this chapter.

In [11]:
def development_summary(reported: cl.Triangle(), paid: cl.Triangle()) -> pd.DataFrame():
    output = pd.DataFrame() # initializing a DataFrame
    output["Reported Claims"] = reported.latest_diagonal.to_frame(origin_as_datetime=False) # using a vector of losses to anchor the exhibit index
    age = reported.development.iloc[::-1] # flipping the age order
    age.index = output.index # forcing the index to match
    output['Age'] = age
    output = output[['Age','Reported Claims']] # reordering the columns
    output ['Paid Claims'] = paid.latest_diagonal.to_frame(origin_as_datetime=False) # adding in paid losses
    reported_cdf = reported.cdf_.T # transposing the CDF
    reported_cdf.index = output.index[::-1] # forcing the index to match
    output["Reported CDF"] = reported_cdf 
    paid_cdf = paid.cdf_.T
    paid_cdf.index = output.index[::-1]
    output["Paid CDF"] = paid_cdf
    output["Reported Ultimate"] = cl.Chainladder().fit(reported).ultimate_.to_frame(origin_as_datetime=False) # using the Chainladder estimator to return the ultimate
    output["Paid Ultimate"] = cl.Chainladder().fit(paid).ultimate_.to_frame(origin_as_datetime=False)
    return output

exhibit = development_summary(reported_selected_pattern,paid_selected_pattern)
display(exhibit)

,Age,Reported Claims,Paid Claims,Reported CDF,Paid CDF,Reported Ultimate,Paid Ultimate
1998,120,47742304.0,47644187.0,1.000000,1.002000,4.774230e+07,4.773948e+07
1999,108,51185767.0,51000534.0,1.000369,1.003870,5.120467e+07,5.119788e+07
2000,96,54837929.0,54533225.0,1.000954,1.006219,5.489024e+07,5.487237e+07
2001,84,56299562.0,55878421.0,1.002540,1.011036,5.644257e+07,5.649507e+07
2002,72,58592712.0,57807215.0,1.005300,1.020604,5.890327e+07,5.899830e+07
2003,60,57565344.0,55930654.0,1.009908,1.039752,5.813573e+07,5.815399e+07
2004,48,56976657.0,53774672.0,1.021554,1.085341,5.820476e+07,5.836386e+07
2005,36,56786410.0,50644994.0,1.049493,1.184218,5.959697e+07,5.997474e+07
2006,24,54641339.0,43606497.0,1.108139,1.404485,6.055018e+07,6.124466e+07
2007,12,48853563.0,27229969.0,1.289977,2.390688,6.301997e+07,6.509837e+07


Unfortunately this does not match the table from the text, due to rounding. We will construct a separate, rounded exhibit to reconcile to the text.

In [38]:
def rounded_development_summary(reported: cl.Triangle(), paid: cl.Triangle()) -> pd.DataFrame():
    output = pd.DataFrame() # initializing a DataFrame
    output["Reported Claims"] = reported.latest_diagonal.to_frame(origin_as_datetime=False) # using a vector of losses to anchor the exhibit index
    age = reported.development.iloc[::-1] # flipping the age order
    age.index = output.index # forcing the index to match
    output['Age'] = age
    output = output[['Age','Reported Claims']] # reordering the columns
    output ['Paid Claims'] = paid.latest_diagonal.to_frame(origin_as_datetime=False) # adding in paid losses
    reported_cdf = reported.ldf_.round(decimals = 3).incr_to_cum().round(decimals = 3).T
    reported_cdf.index = output.index[::-1]
    output["Reported CDF"] = reported_cdf
    paid_cdf = paid.ldf_.round(decimals = 3).incr_to_cum().round(decimals = 3).T
    paid_cdf.index = output.index[::-1]
    output["Paid CDF"] = paid_cdf
    output["Reported Ultimate"] = (output['Reported Claims'] * output["Reported CDF"]).round(decimals = 0) # taking a short cut to calculate the ultimate without using Chainladder 
    output["Paid Ultimate"] = (output['Paid Claims'] * output["Paid CDF"]).round(decimals = 0)
    return output

rounded_exhibit = rounded_development_summary(reported_selected_pattern,paid_selected_pattern)
display(rounded_exhibit[['Reported CDF','Paid CDF','Reported Ultimate','Paid Ultimate']]) # only displaying the rounded columns


,Reported CDF,Paid CDF,Reported Ultimate,Paid Ultimate
1998,1.000,1.002,47742304.0,47739475.0
1999,1.000,1.004,51185767.0,51204536.0
2000,1.001,1.006,54892767.0,54860424.0
2001,1.003,1.011,56468461.0,56493084.0
2002,1.006,1.020,58944268.0,58963359.0
2003,1.011,1.040,58198563.0,58167880.0
2004,1.023,1.085,58287120.0,58345519.0
2005,1.051,1.184,59682517.0,59963673.0
2006,1.110,1.404,60651886.0,61223522.0
2007,1.292,2.390,63118803.0,65079626.0


### Exhibit I, Sheet 4 p109

This is another common report layout for reserving analyses. The manipulation here are more straight-forward that the previous exhibit.

In [39]:
def unpaid_summary(dev_sum: pd.DataFrame()) -> pd.DataFrame():
    output = dev_sum.loc[:,['Reported Claims','Paid Claims','Reported Ultimate','Paid Ultimate']]
    output['Case Outstanding'] = output['Reported Claims'] - output['Paid Claims']
    output['Reported Method IBNR'] = output['Reported Ultimate'] - output['Reported Claims']
    output['Paid Method IBNR'] = output['Paid Ultimate'] - output['Reported Claims']
    output['Reported Method Unpaid'] = output['Reported Method IBNR'] + output['Case Outstanding']
    output['Paid Method Unpaid'] = output['Paid Method IBNR'] + output['Case Outstanding']
    return output

unpaid_exhibit = unpaid_summary(rounded_exhibit)
display(unpaid_exhibit[['Case Outstanding','Reported Method IBNR','Paid Method IBNR','Reported Method Unpaid','Paid Method Unpaid']])

,Case Outstanding,Reported Method IBNR,Paid Method IBNR,Reported Method Unpaid,Paid Method Unpaid
1998,98117.0,0.0,-2829.0,98117.0,95288.0
1999,185233.0,0.0,18769.0,185233.0,204002.0
2000,304704.0,54838.0,22495.0,359542.0,327199.0
2001,421141.0,168899.0,193522.0,590040.0,614663.0
2002,785497.0,351556.0,370647.0,1137053.0,1156144.0
2003,1634690.0,633219.0,602536.0,2267909.0,2237226.0
2004,3201985.0,1310463.0,1368862.0,4512448.0,4570847.0
2005,6141416.0,2896107.0,3177263.0,9037523.0,9318679.0
2006,11034842.0,6010547.0,6582183.0,17045389.0,17617025.0
2007,21623594.0,14265240.0,16226063.0,35888834.0,37849657.0


# Exhibit II - XYZ Insurer

In Exhibit II, Friedland applies the Development method to XYZ Insurer, a private passenger automobile bodily injury liability insurer to illustrate how the method performs in scenarios where due to changing conditions historical claims experience may not be an appropriate predictor of future claims activity.  We reproduce these results in the following cells.

### Exhibit II, Sheet 1 p110

Now that we have walked through an analysis step by step, let's introduce some scaling by streamlining the entire exhibit into single function. 

In [57]:
def dev_exhibit(tri: cl.Triangle, avg_params: dict[str,int], selected_avg: str, tail: float) -> dict[cl.Triangle()]:
    print('PART 1 - Data Triangle')
    print(tri)
    print('PART 2 - Age-to-Age Factors')
    print(tri.age_to_age)
    devs = {}
    print('PART 3 - Average Age-to-Age Factor')
    for k,v in avg_params.items():
        devs[k] = cl.Development(**v).fit_transform(tri)
    def print_ldfs(ldf_dict:dict[cl.Triangle()]):
        print(pd.concat([v.to_frame().rename(index={'(All)':k}) for k,v in ldf_dict.items()]))
        return None
    print_ldfs({k:v.ldf_.round(decimals=3) for k,v in devs.items()})
    devs["Selected"] = cl.TailConstant(tail = tail, projection_period = 0).fit_transform(devs[selected_avg])
    selected = {}
    selected['CDF to Ultimate'] = devs["Selected"].ldf_.round(decimals=3).incr_to_cum().round(decimals=3)
    selected['Percent Reported'] = (1/selected['CDF to Ultimate']).round(decimals=3)
    print('PART 4 - Selected Age-to-Age Factor')
    print_ldfs({'Selected':devs['Selected'].ldf_.round(decimals=3)})
    print_ldfs(selected)
    return devs

import re
tri = cl.load_sample('friedland_xyz_auto_bi')
assumptions_list = ['simple_5','simple_3','simple_2','volume_4','volume_3','volume_2','geometric_3']
assumptions = {x:{'n_periods':int(re.match(r'.+_(.+)', x).group(1)),'average':re.match(r'(.+)_', x).group(1)} for x in assumptions_list}
assumptions['medial 5x1'] = {'n_periods':5, 'average':'simple','drop_high':1, 'drop_low':1}
reported_devs = dev_exhibit(tri['Reported Claims'],assumptions,'volume_2',1)
display(reported_devs)

PART 1 - Data Triangle
          12       24       36       48       60       72       84       96       108      120      132
1998      NaN      NaN  11171.0  12380.0  13216.0  14067.0  14688.0  16366.0  16163.0  15835.0  15822.0
1999      NaN  13255.0  16405.0  19639.0  22473.0  23764.0  25094.0  24795.0  25071.0  25107.0      NaN
2000  15676.0  18749.0  21900.0  27144.0  29488.0  34458.0  36949.0  37505.0  37246.0      NaN      NaN
2001  11827.0  16004.0  21022.0  26578.0  34205.0  37136.0  38541.0  38798.0      NaN      NaN      NaN
2002  12811.0  20370.0  26656.0  37667.0  44414.0  48701.0  48169.0      NaN      NaN      NaN      NaN
2003   9651.0  16995.0  30354.0  40594.0  44231.0  44373.0      NaN      NaN      NaN      NaN      NaN
2004  16995.0  40180.0  58866.0  71707.0  70288.0      NaN      NaN      NaN      NaN      NaN      NaN
2005  28674.0  47432.0  70340.0  70655.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN
2006  27066.0  46783.0  48804.0      NaN 

C:\Users\Shahen Alexanian\Documents\Github\chainladder-python\chainladder\utils\triangle_weight.py:361: UserWarning: Some exclusions have been ignored. At least 1 (use preserve = ...) link ratio(s) is required for development estimation.
  warnings.warn(warning)
C:\Users\Shahen Alexanian\Documents\Github\chainladder-python\chainladder\development\base.py:272: UserWarning: Some exclusions have been ignored. At least 1 (use preserve = ...) link ratio(s) is required for development estimation.
  warnings.warn(warning)


{'simple_5':           12       24       36       48       60       72       84       96       108      120      132
 1998      NaN      NaN  11171.0  12380.0  13216.0  14067.0  14688.0  16366.0  16163.0  15835.0  15822.0
 1999      NaN  13255.0  16405.0  19639.0  22473.0  23764.0  25094.0  24795.0  25071.0  25107.0      NaN
 2000  15676.0  18749.0  21900.0  27144.0  29488.0  34458.0  36949.0  37505.0  37246.0      NaN      NaN
 2001  11827.0  16004.0  21022.0  26578.0  34205.0  37136.0  38541.0  38798.0      NaN      NaN      NaN
 2002  12811.0  20370.0  26656.0  37667.0  44414.0  48701.0  48169.0      NaN      NaN      NaN      NaN
 2003   9651.0  16995.0  30354.0  40594.0  44231.0  44373.0      NaN      NaN      NaN      NaN      NaN
 2004  16995.0  40180.0  58866.0  71707.0  70288.0      NaN      NaN      NaN      NaN      NaN      NaN
 2005  28674.0  47432.0  70340.0  70655.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN
 2006  27066.0  46783.0  48804.0      NaN  

### Exhibit II, Sheet 2 (p. 111)

In [55]:
paid_devs = dev_exhibit(tri['Paid Claims'],assumptions,'volume_2',1.01)

PART 1 - Data Triangle
         12       24       36       48       60       72       84       96       108      120      132
1998     NaN      NaN   6309.0   8521.0  10082.0  11620.0  13242.0  14419.0  15311.0  15764.0  15822.0
1999     NaN   4666.0   9861.0  13971.0  18127.0  22032.0  23511.0  24146.0  24592.0  24817.0      NaN
2000  1302.0   6513.0  12139.0  17828.0  24030.0  28853.0  33222.0  35902.0  36782.0      NaN      NaN
2001  1539.0   5952.0  12319.0  18609.0  24387.0  31090.0  37070.0  38519.0      NaN      NaN      NaN
2002  2318.0   7932.0  13822.0  22095.0  31945.0  40629.0  44437.0      NaN      NaN      NaN      NaN
2003  1743.0   6240.0  12683.0  22892.0  34505.0  39320.0      NaN      NaN      NaN      NaN      NaN
2004  2221.0   9898.0  25950.0  43439.0  52811.0      NaN      NaN      NaN      NaN      NaN      NaN
2005  3043.0  12219.0  27073.0  40026.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN
2006  3531.0  11778.0  22819.0      NaN      NaN  

C:\Users\Shahen Alexanian\Documents\Github\chainladder-python\chainladder\utils\triangle_weight.py:361: UserWarning: Some exclusions have been ignored. At least 1 (use preserve = ...) link ratio(s) is required for development estimation.
  warnings.warn(warning)
C:\Users\Shahen Alexanian\Documents\Github\chainladder-python\chainladder\development\base.py:272: UserWarning: Some exclusions have been ignored. At least 1 (use preserve = ...) link ratio(s) is required for development estimation.
  warnings.warn(warning)


### Exhibit II, Sheet 3 (p. 112)

In [56]:
exhibit = rounded_development_summary(reported_devs["Selected"],paid_devs["Selected"])
display(exhibit)

,Age,Reported Claims,Paid Claims,Reported CDF,Paid CDF,Reported Ultimate,Paid Ultimate
1998,132,15822.0,15822.0,1.000,1.010,15822.0,15980.0
1999,120,25107.0,24817.0,0.999,1.014,25082.0,25164.0
2000,108,37246.0,36782.0,0.992,1.031,36948.0,37922.0
2001,96,38798.0,38519.0,0.992,1.054,38488.0,40599.0
2002,84,48169.0,44437.0,1.003,1.116,48314.0,49592.0
2003,72,44373.0,39320.0,1.013,1.268,44950.0,49858.0
2004,60,70288.0,52811.0,1.064,1.525,74786.0,80537.0
2005,48,70655.0,40026.0,1.085,2.007,76661.0,80332.0
2006,36,48804.0,22819.0,1.196,3.160,58370.0,72108.0
2007,24,31732.0,11865.0,1.512,6.569,47979.0,77941.0


### Exhibit II, Sheet 4 p113

In [58]:
unpaid_exhibit = unpaid_summary(exhibit)
display(unpaid_exhibit[['Case Outstanding','Reported Method IBNR','Paid Method IBNR','Reported Method Unpaid','Paid Method Unpaid']])

,Case Outstanding,Reported Method IBNR,Paid Method IBNR,Reported Method Unpaid,Paid Method Unpaid
1998,0.0,0.0,158.0,0.0,158.0
1999,290.0,-25.0,57.0,265.0,347.0
2000,464.0,-298.0,676.0,166.0,1140.0
2001,279.0,-310.0,1801.0,-31.0,2080.0
2002,3732.0,145.0,1423.0,3877.0,5155.0
2003,5053.0,577.0,5485.0,5630.0,10538.0
2004,17477.0,4498.0,10249.0,21975.0,27726.0
2005,30629.0,6006.0,9677.0,36635.0,40306.0
2006,25985.0,9566.0,23304.0,35551.0,49289.0
2007,19867.0,16247.0,46209.0,36114.0,66076.0


### Persisting the Selected Estimators

The selected reported and paid development estimators are reused in later chapters, such as the Bornhuetter-Ferguson technique in Chapter 9. Rather than refitting them, we persist the fitted estimators with ``to_json`` so they can be recalled directly. JSON is used rather than a pickle so the saved estimators load reliably across package and dependency versions.

In [59]:
import os
data_dir = os.path.join(os.path.dirname(cl.__file__), "utils", "data")

with open(os.path.join(data_dir, "friedland_ch7_xyz_reported.json"), "w") as f:
    _ = f.write(reported_devs["Selected"].to_json())

with open(os.path.join(data_dir, "friedland_ch7_xyz_paid.json"), "w") as f:
    _ = f.write(paid_devs["Selected"].to_json())

## Exhibit III — U.S. Industry P.P. Auto (Impact of Changing Conditions)

In order to illustrate the assumptions and corresponding limitations of the Development method, Exhibit III applies the method to the **U.S. Private Passenger Industry Auto** (USPP Auto) example under four different scenarios:

1. stable claim ratios and no change in case outstanding strength (steady state);
1. increasing claim ratios but no change in case outstanding strength;
1. stable claim ratios but increasing case outstanding strength; and
1. increasing claim ratios and increasing case outstanding strength.

The key assumptions of this study are as follows:

- the earned premium for the first year (1999) is assumed to be $1 million with a 5% annual premium trend;
- (discuss assumed claim ratios)

Since the USPP Auto dataset is contained in the `chainladder` package, we simply load it into a triangle as follows:

In [6]:
triangles = cl.load_sample("friedland_uspp")

In [ ]:
def dev_exhibit(tri: cl.Triangle, avg_params: dict[str,int], selected_avg: str, tail: float) -> dict[cl.Triangle()]:
    display('')
    display(HTML("""
    <h2 style='text-align:left;'>
    PART 1 - Data Triangle
    </h2>
    """))
    display(tri)
    display(HTML("""
    <h2 style='text-align:left;'>
    PART 2 - Age-to-Age Factors
    </h2>
    """))
    age_to_age_df = tri.age_to_age.to_frame(origin_as_datetime=False)
    display(
        age_to_age_df.style.format(precision=3, na_rep="")
    )
    devs = {}
    display(HTML("""
    <h2 style='text-align:left;'>
    PART 3 - Average Age-to-Age Factor
    </h2>
    """))
    for k,v in avg_params.items():
        devs[k] = cl.Development(**v).fit_transform(tri)
    def print_ldfs(ldf_dict:dict[cl.Triangle()]):
        with pd.option_context("display.float_format", "{:.3f}".format):
            display(pd.concat([v.to_frame().rename(index={'(All)':k}) for k,v in ldf_dict.items()]))
        return None
    print_ldfs({k:v.ldf_.round(decimals=3) for k,v in devs.items()})
    devs["Selected"] = cl.TailConstant(tail = tail, projection_period = 0).fit_transform(devs[selected_avg])
    selected = {}
    selected['CDF to Ultimate'] = devs["Selected"].ldf_.round(decimals=3).incr_to_cum().round(decimals=3)
    selected['Percent Reported'] = (1/selected['CDF to Ultimate']).round(decimals=3)
    display(HTML("""
    <h2 style='text-align:left;'>
    PART 4 - Selected Age-to-Age Factor
    </h2>
    """))
    print_ldfs({'Selected':devs['Selected'].ldf_.round(decimals=3)})
    print_ldfs(selected)
    return devs

In [10]:
#| hide-input

def ex3sht1(
          tri: cl.Triangle,
          dev_input: dict,
          tail_input: dict
) -> tuple:
    
    def format_col(x, reverse = False):
            if reverse is True:
                return x.to_frame().squeeze()[::-1].reset_index(drop=True)
            return x.to_frame().squeeze().reset_index(drop=True)

    def col_diff(x, y):
        return x.squeeze() - y.squeeze()

    def col_div(x, y):
        return x.squeeze() / y.squeeze()

    def col_mult(x, y):
        return x.squeeze() * y.squeeze()

    title = ("Exhibit III, Sheet 1, Summary of Earned Premium and Claim Ratio Assumptions and Actual IBNR")
    #scenario = "Steady State"

    dev = cl.Development(**dev_input).fit(tri)
    model = cl.Chainladder().fit(dev.transform(tri))
    tail = cl.TailConstant(**tail_input).fit_transform(dev.transform(tri))
    ult = model.ultimate_
    ibnr = model.ibnr_

    col1 = pd.DataFrame({"Accident Year" : format_col(tri.origin.astype(str))})
    col2 = pd.DataFrame({"Earned Premium" : format_col(tri.latest_diagonal.loc["Steady State", "Earned Premium"])})
    col4 = pd.DataFrame({"Ult. Claims" : format_col(ult.loc["Steady State", "Reported Claims"])})
    col3 = pd.DataFrame({"Ult. Claim Ratio" : col_div(col4, col2)})
    col5 = pd.DataFrame({"Rep. Claims 12/31/08" : format_col(tri.latest_diagonal.loc["Steady State", "Reported Claims"])})
    col6 = pd.DataFrame({"Actual IBNR" : col_diff(col4, col5)})
    col8 = pd.DataFrame({"Ult. Claims" : format_col(ult.loc["Increasing Claim", "Reported Claims"])})
    col9 = pd.DataFrame({"Rep. Claims 12/31/08" : format_col(tri.latest_diagonal.loc["Increasing Claim", "Reported Claims"])})
    col7 = pd.DataFrame({"Ult. Claim Ratio" : col_div(col8, col2)})
    col10 = pd.DataFrame({"Actual IBNR" : col_diff(col8, col9)})
    col11 = pd.DataFrame({"Accident Year" : tri.origin.astype(str)})
    col12 = pd.DataFrame({"Earned Premium" : format_col(tri.latest_diagonal.loc["Steady State", "Earned Premium"])})
    col14 = pd.DataFrame({"Ult. Claims" : format_col(ult.loc["Steady State", "Reported Claims"])})
    col13 = pd.DataFrame({"Ult. Claim Ratio" : col_div(col14, col12)})
    col15 = pd.DataFrame({"Rep. Claims 12/31/08" : format_col(tri.latest_diagonal.loc["Increasing Case", "Reported Claims"])})
    col16 = pd.DataFrame({"Actual IBNR" : col_diff(col14, col15)})
    col18 = pd.DataFrame({"Ult. Claims" : format_col(ult.loc["Increasing Claim", "Reported Claims"])})
    col17 = pd.DataFrame({"Ult. Claim Ratio" : col_div(col18, col12)})
    col19 = pd.DataFrame({"Rep. Claims 12/31/08" : format_col(tri.latest_diagonal.loc["Increasing Claim Case", "Reported Claims"])})
    col20 = pd.DataFrame({"Actual IBNR" : col_diff(col18, col19)})

    cols = [col1, col2, col3, col4, col5, col6]
    df1 = pd.concat([col1, col2], axis=1)
    df2 = pd.concat([col3, col4, col5, col6], axis=1)
    df3 = pd.concat([col7, col8, col9, col10], axis=1)
    df4 = pd.concat([col11, col12], axis=1)
    df5 = pd.concat([col13, col14, col15, col16], axis=1)
    df6 = pd.concat([col17, col18, col19, col20], axis=1)

    dfs_upper = [df1, df2, df3]
    dfs_lower = [df4, df5, df6]

    results_upper = pd.concat(dfs_upper, axis=1, keys=["", "Steady State", "Increasing Claim Ratios"])
    results_upper = results_upper.set_index(("", "Accident Year")) #since its multi-index, I needed to use the tuple to designate the index col
    results_upper.index.name = "Accident Year" 
    results_upper.loc["Total"] = results_upper.sum()

    #remove the totals from the ratio columns (where totals do not make sense)
    results_upper.loc["Total", ("Steady State", "Ult. Claim Ratio")] = np.nan
    results_upper.loc["Total", ("Increasing Claim Ratios", "Ult. Claim Ratio")] = np.nan

    results_lower = pd.concat(dfs_lower, axis=1, keys=["", "Increasing Case Outstanding Strength", "Increasing Claim Ratios and Case Outstanding Strength"])
    results_lower = results_lower.set_index(("", "Accident Year")) #since its multi-index, I needed to use the tuple to designate the index col
    results_lower.index.name = "Accident Year" 
    results_lower.loc["Total"] = results_lower.sum()

    #remove the totals from the ratio columns (where totals do not make sense)
    results_lower.loc["Total", ("Increasing Case Outstanding Strength", "Ult. Claim Ratio")] = np.nan
    results_lower.loc["Total", ("Increasing Claim Ratios and Case Outstanding Strength", "Ult. Claim Ratio")] = np.nan

    display(HTML("""
    <h2 style='text-align:center;'>
    Exhibit III Sheet 1: Summary of Earned Premium and Claim Ratio Assumptions and Actual IBNR
    </h2>
    """))
    display(results_upper)
    display(results_lower)

    return (results_upper, results_lower)

## Exhibit III — U.S. Private-Passenger Industry Auto - Impact of Changing Conditions

### Exhibit III, Sheet 1 - Summary of Earned Premium and Claim Ratio Assumptions and Actual IBNR

We first reproduce Exhibit III, Sheet 1 of Chapter 7 of Friedland (p. 114), which summarizes the earned premium and claim ratio assumptions under the four scenarios:

In [13]:
dev_input = {
    "average" : "volume",
    "n_periods" : 5
}

tail_input = {
    "tail" : 1.0,
    "projection_period" : -1
}

res_up, res_low = ex3sht1(triangles, dev_input, tail_input)

Steady State                                              Increasing Claim Ratios                                             
              Earned Premium Ult. Claim Ratio Ult. Claims Rep. Claims 12/31/08 Actual IBNR        Ult. Claim Ratio Ult. Claims Rep. Claims 12/31/08 Actual IBNR
Accident Year                                                                                                                                                  
1999               1,000,000                1     700,000              700,000           0                       1     700,000              700,000           0
2000               1,050,000                1     735,000              735,000           0                       1     735,000              735,000           0
2001               1,102,500                1     771,750              771,750           0                       1     771,750              771,750           0
2002               1,157,625                1     810,338              810,338           0                       1     810,338              810,338           0
2003               1,215,506                1     850,855              842,346       8,509                       1     850,855              842,346       8,509
2004               1,276,282                1     893,397              884,463       8,934                       1   1,021,025            1,010,815      10,210
2005               1,340,096                1     938,068              919,306      18,762                       1   1,139,082            1,116,300      22,782
2006               1,407,100                1     984,970              935,722      49,248                       1   1,266,391            1,203,071      63,320
2007               1,477,455                1   1,034,219              930,797     103,422                       1   1,403,582            1,263,224     140,358
2008               1,551,328                1   1,085,930              836,166     249,764                       1   1,551,328            1,194,523     356,805
Total             12,577,893              NaN   8,804,527            8,365,888     438,639                     NaN  10,249,351            9,647,367     601,984

Increasing Case Outstanding Strength                                              Increasing Claim Ratios and Case Outstanding Strength                                             
              Earned Premium                     Ult. Claim Ratio Ult. Claims Rep. Claims 12/31/08 Actual IBNR                                      Ult. Claim Ratio Ult. Claims Rep. Claims 12/31/08 Actual IBNR
Accident Year                                                                                                                                                                                                    
1999               1,000,000                                    1     700,000              700,000           0                                                  1        700,000              700,000           0
2000               1,050,000                                    1     735,000              735,000           0                                                  1        735,000              735,000           0
2001               1,102,500                                    1     771,750              771,750           0                                                  1        771,750              771,750           0
2002               1,157,625                                    1     810,338              810,338           0                                                  1        810,338              810,338           0
2003               1,215,506                                    1     850,855              842,346       8,509                                                  1        850,855              842,346       8,509
2004               1,276,282                                    1     893,397              884,463       8,934                                                  1      1,021,025            1,010,815      10,210
2005               1,340,096                                    1     938,068              933,377       4,691                                                  1      1,139,082            1,133,386       5,696
2006               1,407,100                                    1     984,970              962,808      22,162                                                  1      1,266,391            1,237,897      28,494
2007               1,477,455                                    1   1,034,219              979,922      54,297                                                  1      1,403,582            1,329,895      73,687
2008               1,551,328                                    1   1,085,930              931,185     154,745                                                  1      1,551,328            1,330,264     221,064
Total             12,577,893                                  NaN   8,804,527            8,551,189     253,338                                                NaN     10,249,351            9,901,691     347,660

### Exhibit III, Sheet 2 - Sheet 2: USPP Auto Steady-State - Reported Claims

We then generate Sheet 2, Parts 1-4 which contain the data triangle and age-to-age factors for USPP reported claims under the steady-state scenario (p. 115).

In [35]:
_ = dev_exhibit(triangles.loc["Steady State", "Reported Claims"],avg_params={"volume_5": {'n_periods': 5, 'average': 'volume'}},selected_avg='volume_5',tail=1)

''

,12,24,36,48,60,72,84,96,108,120
1999,"539,000","630,000","665,000","686,000","693,000","693,000","700,000","700,000","700,000","700,000"
2000,"565,950","661,500","698,250","720,300","727,650","727,650","735,000","735,000","735,000",
2001,"594,248","694,575","733,163","756,315","764,033","764,033","771,750","771,750",,
2002,"623,960","729,304","769,821","794,131","802,234","802,234","810,338",,,
2003,"655,158","765,769","808,312","833,837","842,346","842,346",,,,
2004,"687,916","804,057","848,727","875,529","884,463",,,,,
2005,"722,312","844,260","891,164","919,306",,,,,,
2006,"758,427","886,473","935,722",,,,,,,
2007,"796,348","930,797",,,,,,,,
2008,"836,166",,,,,,,,,


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
1999,1.169,1.056,1.032,1.010,1.000,1.010,1.000,1.000,1.000
2000,1.169,1.056,1.032,1.010,1.000,1.010,1.000,1.000,
2001,1.169,1.056,1.032,1.010,1.000,1.010,1.000,,
2002,1.169,1.056,1.032,1.010,1.000,1.010,,,
2003,1.169,1.056,1.032,1.010,1.000,,,,
2004,1.169,1.056,1.032,1.010,,,,,
2005,1.169,1.056,1.032,,,,,,
2006,1.169,1.056,,,,,,,
2007,1.169,,,,,,,,


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
volume_5,1.169,1.056,1.032,1.010,1.000,1.010,1.000,1.000,1.000


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120,120-132
Selected,1.169,1.056,1.032,1.010,1.000,1.010,1.000,1.000,1.000,1.000


,12-Ult,24-Ult,36-Ult,48-Ult,60-Ult,72-Ult,84-Ult,96-Ult,108-Ult,120-Ult
CDF to Ultimate,1.300,1.112,1.053,1.020,1.010,1.010,1.000,1.000,1.000,1.000
Percent Reported,0.769,0.899,0.950,0.980,0.990,0.990,1.000,1.000,1.000,1.000


### Exhibit III, Sheet 3: USPP Auto Steady-State - Paid Claims

We then generate Sheet 3, Parts 1-4 which contain the data triangle and age-to-age factors for USPP paid claims under the steady-state scenario (p. 116).

In [36]:
_ = dev_exhibit(triangles.loc["Steady State", "Paid Claims"],avg_params={"volume_5": {'n_periods': 5, 'average': 'volume'}},selected_avg='volume_5',tail=1)

''

,12,24,36,48,60,72,84,96,108,120
1999,"294,000","497,000","588,000","644,000","672,000","686,000","693,000","693,000","700,000","700,000"
2000,"308,700","521,850","617,400","676,200","705,600","720,300","727,650","727,650","735,000",
2001,"324,135","547,943","648,270","710,010","740,880","756,315","764,033","764,033",,
2002,"340,342","575,340","680,684","745,511","777,924","794,131","802,234",,,
2003,"357,359","604,107","714,718","782,786","816,820","833,837",,,,
2004,"375,227","634,312","750,454","821,925","857,661",,,,,
2005,"393,988","666,028","787,976","863,022",,,,,,
2006,"413,688","699,329","827,375",,,,,,,
2007,"434,372","734,295",,,,,,,,
2008,"456,090",,,,,,,,,


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
1999,1.690,1.183,1.095,1.043,1.021,1.010,1.000,1.010,1.000
2000,1.690,1.183,1.095,1.043,1.021,1.010,1.000,1.010,
2001,1.690,1.183,1.095,1.043,1.021,1.010,1.000,,
2002,1.690,1.183,1.095,1.043,1.021,1.010,,,
2003,1.690,1.183,1.095,1.043,1.021,,,,
2004,1.690,1.183,1.095,1.043,,,,,
2005,1.690,1.183,1.095,,,,,,
2006,1.690,1.183,,,,,,,
2007,1.690,,,,,,,,


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
volume_5,1.690,1.183,1.095,1.043,1.021,1.010,1.000,1.010,1.000


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120,120-132
Selected,1.690,1.183,1.095,1.043,1.021,1.010,1.000,1.010,1.000,1.000


,12-Ult,24-Ult,36-Ult,48-Ult,60-Ult,72-Ult,84-Ult,96-Ult,108-Ult,120-Ult
CDF to Ultimate,2.378,1.407,1.190,1.086,1.042,1.020,1.010,1.010,1.000,1.000
Percent Reported,0.421,0.711,0.840,0.921,0.960,0.980,0.990,0.990,1.000,1.000


### Exhibit III, Sheet 4: USPP Auto Increasing Claim Ratios - Reported Claims

We then generate Sheet 4, Parts 1-4 which contain the data triangle and age-to-age factors for USPP reported claims under the increasing claim ratio scenario (p. 117).

In [37]:
_ = dev_exhibit(triangles.loc["Increasing Claim", "Reported Claims"],avg_params={"volume_5": {'n_periods': 5, 'average': 'volume'}},selected_avg='volume_5',tail=1)

''

,12,24,36,48,60,72,84,96,108,120
1999,"539,000","630,000","665,000","686,000","693,000","693,000","700,000","700,000","700,000","700,000"
2000,"565,950","661,500","698,250","720,300","727,650","727,650","735,000","735,000","735,000",
2001,"594,248","694,575","733,163","756,315","764,033","764,033","771,750","771,750",,
2002,"623,960","729,304","769,821","794,131","802,234","802,234","810,338",,,
2003,"655,158","765,769","808,312","833,837","842,346","842,346",,,,
2004,"786,189","918,923","969,974","1,000,605","1,010,815",,,,,
2005,"877,093","1,025,173","1,082,127","1,116,300",,,,,,
2006,"975,121","1,139,751","1,203,071",,,,,,,
2007,"1,080,759","1,263,224",,,,,,,,
2008,"1,194,523",,,,,,,,,


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
1999,1.169,1.056,1.032,1.010,1.000,1.010,1.000,1.000,1.000
2000,1.169,1.056,1.032,1.010,1.000,1.010,1.000,1.000,
2001,1.169,1.056,1.032,1.010,1.000,1.010,1.000,,
2002,1.169,1.056,1.032,1.010,1.000,1.010,,,
2003,1.169,1.056,1.032,1.010,1.000,,,,
2004,1.169,1.056,1.032,1.010,,,,,
2005,1.169,1.056,1.032,,,,,,
2006,1.169,1.056,,,,,,,
2007,1.169,,,,,,,,


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
volume_5,1.169,1.056,1.032,1.010,1.000,1.010,1.000,1.000,1.000


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120,120-132
Selected,1.169,1.056,1.032,1.010,1.000,1.010,1.000,1.000,1.000,1.000


,12-Ult,24-Ult,36-Ult,48-Ult,60-Ult,72-Ult,84-Ult,96-Ult,108-Ult,120-Ult
CDF to Ultimate,1.300,1.112,1.053,1.020,1.010,1.010,1.000,1.000,1.000,1.000
Percent Reported,0.769,0.899,0.950,0.980,0.990,0.990,1.000,1.000,1.000,1.000


### Exhibit III, Sheet 5: USPP Auto Increasing Claim Ratios - Paid Claims

We then generate Sheet 5, Parts 1-4 which contain the data triangle and age-to-age factors for USPP paid claims under the increasing claim ratio scenario (p. 118).

In [38]:
_ = dev_exhibit(triangles.loc["Increasing Claim", "Paid Claims"],avg_params={"volume_5": {'n_periods': 5, 'average': 'volume'}},selected_avg='volume_5',tail=1)

''

,12,24,36,48,60,72,84,96,108,120
1999,"294,000","497,000","588,000","644,000","672,000","686,000","693,000","693,000","700,000","700,000"
2000,"308,700","521,850","617,400","676,200","705,600","720,300","727,650","727,650","735,000",
2001,"324,135","547,943","648,270","710,010","740,880","756,315","764,033","764,033",,
2002,"340,342","575,340","680,684","745,511","777,924","794,131","802,234",,,
2003,"357,359","604,107","714,718","782,786","816,820","833,837",,,,
2004,"428,831","724,928","857,661","939,343","980,184",,,,,
2005,"478,414","808,748","956,828","1,047,955",,,,,,
2006,"531,884","899,137","1,063,768",,,,,,,
2007,"589,505","996,544",,,,,,,,
2008,"651,558",,,,,,,,,


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
1999,1.690,1.183,1.095,1.043,1.021,1.010,1.000,1.010,1.000
2000,1.690,1.183,1.095,1.043,1.021,1.010,1.000,1.010,
2001,1.690,1.183,1.095,1.043,1.021,1.010,1.000,,
2002,1.690,1.183,1.095,1.043,1.021,1.010,,,
2003,1.690,1.183,1.095,1.043,1.021,,,,
2004,1.690,1.183,1.095,1.043,,,,,
2005,1.690,1.183,1.095,,,,,,
2006,1.690,1.183,,,,,,,
2007,1.690,,,,,,,,


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
volume_5,1.690,1.183,1.095,1.043,1.021,1.010,1.000,1.010,1.000


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120,120-132
Selected,1.690,1.183,1.095,1.043,1.021,1.010,1.000,1.010,1.000,1.000


,12-Ult,24-Ult,36-Ult,48-Ult,60-Ult,72-Ult,84-Ult,96-Ult,108-Ult,120-Ult
CDF to Ultimate,2.378,1.407,1.190,1.086,1.042,1.020,1.010,1.010,1.000,1.000
Percent Reported,0.421,0.711,0.840,0.921,0.960,0.980,0.990,0.990,1.000,1.000


### Exhibit III, Sheet 6: USPP Auto Increasing Case Outstanding Strength - Reported Claims

We then generate Sheet 6, Parts 1-4 which contain the data triangle and age-to-age factors for USPP reported claims under the increasing case outstanding strength scenario (p. 119).

In [39]:
_ = dev_exhibit(triangles.loc["Increasing Case", "Reported Claims"],avg_params={"volume_5": {'n_periods': 5, 'average': 'volume'}},selected_avg='volume_5',tail=1)

''

,12,24,36,48,60,72,84,96,108,120
1999,"539,000","630,000","665,000","686,000","693,000","693,000","700,000","700,000","700,000","700,000"
2000,"565,950","661,500","698,250","720,300","727,650","727,650","735,000","735,000","735,000",
2001,"594,248","694,575","733,163","756,315","764,033","764,033","771,750","771,750",,
2002,"623,960","729,304","769,821","794,131","802,234","802,234","810,338",,,
2003,"655,158","765,769","808,312","833,837","842,346","842,346",,,,
2004,"687,916","804,057","848,727","878,745","884,463",,,,,
2005,"722,312","844,260","897,355","933,377",,,,,,
2006,"758,427","897,702","962,808",,,,,,,
2007,"818,067","979,922",,,,,,,,
2008,"931,185",,,,,,,,,


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
1999,1.169,1.056,1.032,1.010,1.000,1.010,1.000,1.000,1.000
2000,1.169,1.056,1.032,1.010,1.000,1.010,1.000,1.000,
2001,1.169,1.056,1.032,1.010,1.000,1.010,1.000,,
2002,1.169,1.056,1.032,1.010,1.000,1.010,,,
2003,1.169,1.056,1.032,1.010,1.000,,,,
2004,1.169,1.056,1.035,1.007,,,,,
2005,1.169,1.063,1.040,,,,,,
2006,1.184,1.073,,,,,,,
2007,1.198,,,,,,,,


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
volume_5,1.178,1.061,1.034,1.009,1.000,1.010,1.000,1.000,1.000


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120,120-132
Selected,1.178,1.061,1.034,1.009,1.000,1.010,1.000,1.000,1.000,1.000


,12-Ult,24-Ult,36-Ult,48-Ult,60-Ult,72-Ult,84-Ult,96-Ult,108-Ult,120-Ult
CDF to Ultimate,1.317,1.118,1.054,1.019,1.010,1.010,1.000,1.000,1.000,1.000
Percent Reported,0.759,0.894,0.949,0.981,0.990,0.990,1.000,1.000,1.000,1.000


### Exhibit III, Sheet 7: USPP Auto Increasing Case Outstanding Strength - Paid Claims

We then generate Sheet 7, Parts 1-4 which contain the data triangle and age-to-age factors for USPP paid claims under the increasing case outstanding strength scenario (p. 120).

In [40]:
_ = dev_exhibit(triangles.loc["Increasing Case", "Paid Claims"],avg_params={"volume_5": {'n_periods': 5, 'average': 'volume'}},selected_avg='volume_5',tail=1)

''

,12,24,36,48,60,72,84,96,108,120
1999,"294,000","497,000","588,000","644,000","672,000","686,000","693,000","693,000","700,000","700,000"
2000,"308,700","521,850","617,400","676,200","705,600","720,300","727,650","727,650","735,000",
2001,"324,135","547,943","648,270","710,010","740,880","756,315","764,033","764,033",,
2002,"340,342","575,340","680,684","745,511","777,924","794,131","802,234",,,
2003,"357,359","604,107","714,718","782,786","816,820","833,837",,,,
2004,"375,227","634,312","750,454","821,925","857,661",,,,,
2005,"393,988","666,028","787,976","863,022",,,,,,
2006,"413,688","699,329","827,375",,,,,,,
2007,"434,372","734,295",,,,,,,,
2008,"456,090",,,,,,,,,


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
1999,1.690,1.183,1.095,1.043,1.021,1.010,1.000,1.010,1.000
2000,1.690,1.183,1.095,1.043,1.021,1.010,1.000,1.010,
2001,1.690,1.183,1.095,1.043,1.021,1.010,1.000,,
2002,1.690,1.183,1.095,1.043,1.021,1.010,,,
2003,1.690,1.183,1.095,1.043,1.021,,,,
2004,1.690,1.183,1.095,1.043,,,,,
2005,1.690,1.183,1.095,,,,,,
2006,1.690,1.183,,,,,,,
2007,1.690,,,,,,,,


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
volume_5,1.690,1.183,1.095,1.043,1.021,1.010,1.000,1.010,1.000


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120,120-132
Selected,1.690,1.183,1.095,1.043,1.021,1.010,1.000,1.010,1.000,1.000


,12-Ult,24-Ult,36-Ult,48-Ult,60-Ult,72-Ult,84-Ult,96-Ult,108-Ult,120-Ult
CDF to Ultimate,2.378,1.407,1.190,1.086,1.042,1.020,1.010,1.010,1.000,1.000
Percent Reported,0.421,0.711,0.840,0.921,0.960,0.980,0.990,0.990,1.000,1.000


### Exhibit III, Sheet 8: USPP Auto Increasing Claim Ratios and Increasing Case Outstanding Strength - Reported Claims

We then generate Sheet 8, Parts 1-4 which contain the data triangle and age-to-age factors for USPP reported claims under the increasing claims ratio / increasing case outstanding strength scenario (p. 121).

In [41]:
_ = dev_exhibit(triangles.loc["Increasing Claim Case", "Reported Claims"],avg_params={"volume_5": {'n_periods': 5, 'average': 'volume'}},selected_avg='volume_5',tail=1)

''

,12,24,36,48,60,72,84,96,108,120
1999,"539,000","630,000","665,000","686,000","693,000","693,000","700,000","700,000","700,000","700,000"
2000,"565,950","661,500","698,250","720,300","727,650","727,650","735,000","735,000","735,000",
2001,"594,248","694,575","733,163","756,315","764,033","764,033","771,750","771,750",,
2002,"623,960","729,304","769,821","794,131","802,234","802,234","810,338",,,
2003,"655,158","765,769","808,312","833,837","842,346","842,346",,,,
2004,"786,189","918,923","969,974","1,004,280","1,010,815",,,,,
2005,"877,093","1,025,173","1,089,645","1,133,386",,,,,,
2006,"975,121","1,154,188","1,237,897",,,,,,,
2007,"1,110,234","1,329,895",,,,,,,,
2008,"1,330,264",,,,,,,,,


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
1999,1.169,1.056,1.032,1.010,1.000,1.010,1.000,1.000,1.000
2000,1.169,1.056,1.032,1.010,1.000,1.010,1.000,1.000,
2001,1.169,1.056,1.032,1.010,1.000,1.010,1.000,,
2002,1.169,1.056,1.032,1.010,1.000,1.010,,,
2003,1.169,1.056,1.032,1.010,1.000,,,,
2004,1.169,1.056,1.035,1.007,,,,,
2005,1.169,1.063,1.040,,,,,,
2006,1.184,1.073,,,,,,,
2007,1.198,,,,,,,,


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
volume_5,1.179,1.061,1.035,1.009,1.000,1.010,1.000,1.000,1.000


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120,120-132
Selected,1.179,1.061,1.035,1.009,1.000,1.010,1.000,1.000,1.000,1.000


,12-Ult,24-Ult,36-Ult,48-Ult,60-Ult,72-Ult,84-Ult,96-Ult,108-Ult,120-Ult
CDF to Ultimate,1.319,1.119,1.055,1.019,1.010,1.010,1.000,1.000,1.000,1.000
Percent Reported,0.758,0.894,0.948,0.981,0.990,0.990,1.000,1.000,1.000,1.000


### Exhibit III, Sheet 9: USPP Auto Increasing Claim Ratios and Increasing Case Outstanding Strength - Paid Claims

We then generate Sheet 9, Parts 1-4 which contain the data triangle and age-to-age factors for USPP paid claims under the increasing claims ratio / increasing case outstanding strength scenario (p. 122):

In [42]:
_ = dev_exhibit(triangles.loc["Increasing Claim Case", "Paid Claims"],avg_params={"volume_5": {'n_periods': 5, 'average': 'volume'}},selected_avg='volume_5',tail=1)

''

,12,24,36,48,60,72,84,96,108,120
1999,"294,000","497,000","588,000","644,000","672,000","686,000","693,000","693,000","700,000","700,000"
2000,"308,700","521,850","617,400","676,200","705,600","720,300","727,650","727,650","735,000",
2001,"324,135","547,943","648,270","710,010","740,880","756,315","764,033","764,033",,
2002,"340,342","575,340","680,684","745,511","777,924","794,131","802,234",,,
2003,"357,359","604,107","714,718","782,786","816,820","833,837",,,,
2004,"428,831","724,928","857,661","939,343","980,184",,,,,
2005,"478,414","808,748","956,828","1,047,955",,,,,,
2006,"531,884","899,137","1,063,768",,,,,,,
2007,"589,505","996,544",,,,,,,,
2008,"651,558",,,,,,,,,


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
1999,1.690,1.183,1.095,1.043,1.021,1.010,1.000,1.010,1.000
2000,1.690,1.183,1.095,1.043,1.021,1.010,1.000,1.010,
2001,1.690,1.183,1.095,1.043,1.021,1.010,1.000,,
2002,1.690,1.183,1.095,1.043,1.021,1.010,,,
2003,1.690,1.183,1.095,1.043,1.021,,,,
2004,1.690,1.183,1.095,1.043,,,,,
2005,1.690,1.183,1.095,,,,,,
2006,1.690,1.183,,,,,,,
2007,1.690,,,,,,,,


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120
volume_5,1.690,1.183,1.095,1.043,1.021,1.010,1.000,1.010,1.000


,12-24,24-36,36-48,48-60,60-72,72-84,84-96,96-108,108-120,120-132
Selected,1.690,1.183,1.095,1.043,1.021,1.010,1.000,1.010,1.000,1.000


,12-Ult,24-Ult,36-Ult,48-Ult,60-Ult,72-Ult,84-Ult,96-Ult,108-Ult,120-Ult
CDF to Ultimate,2.378,1.407,1.190,1.086,1.042,1.020,1.010,1.010,1.000,1.000
Percent Reported,0.421,0.711,0.840,0.921,0.960,0.980,0.990,0.990,1.000,1.000


In [ ]:
#| hide-input

def Ex10Sht10(scenario, tr):

    def format_col(x, reverse = False):
        if reverse is True:
            return x.to_frame().squeeze()[::-1].reset_index(drop=True)
        return x.to_frame().squeeze().reset_index(drop=True)

    def col_diff(x, y):
        return x.squeeze() - y.squeeze()

    def col_mult(x, y):
        return x.squeeze() * y.squeeze()

    tri = tr.copy()

    dev_input = {
        "average" : "volume",
        "n_periods" : 5
    }

    tail_input = {
        "tail" : 1.0,
        "projection_period" : -1
    }

    dev = cl.Development(**dev_input)
    tail = cl.TailConstant(**tail_input)
    tri = tail.fit_transform(dev.fit_transform(tri))

    actual_scenario = (
    "Steady State"
    if scenario in ["Steady State", "Increasing Case"]
    else "Increasing Claim"
    )

    cdf = format_col(tri.cdf_.loc[actual_scenario]["Reported Claims"], reverse=True)
    ult = format_col(tri.loc[actual_scenario]["Reported Claims"].latest_diagonal)
    
    col1 = pd.DataFrame({
        "Accident Year" : format_col(tr.origin)
        })

    col2 = pd.DataFrame({
        "Age of Accident Year at 12/31/08" : format_col(tr.development, reverse=True)

    })

    col3 = pd.DataFrame({
        "Claims at 12/31/2008 - Reported" : format_col(tri.loc[scenario]["Reported Claims"].latest_diagonal)

    })

    col4 = pd.DataFrame({
        "Claims at 12/31/2008 - Paid" : format_col(tri.loc[scenario]["Paid Claims"].latest_diagonal)

    })

    col5 = pd.DataFrame({
        "Case Outstanding" : col_diff(col3, col4)

    })

    col6 = pd.DataFrame({
        "CDF to Ult. - Reported" : format_col(tri.cdf_.loc[scenario]["Reported Claims"], reverse=True)

    })

    col7 = pd.DataFrame({
        "CDF to Ult. - Paid" : format_col(tri.cdf_.loc[scenario]["Paid Claims"], reverse=True)
    })

    col8 = pd.DataFrame({
        "Projected Ult. Claims Using Dev. Method - Reported" : col_mult(col3, col6)
    })

    col9 = pd.DataFrame({
        "Projected Ult. Claims Using Dev. Method - Paid" : col_mult(col4, col7)
    })

    col10 = pd.DataFrame({
        "Estimated IBNR Using Dev. Method - Reported" : col_diff(col8, col3)
    })

    col11 = pd.DataFrame({
        "Estimated IBNR Using Dev. Method - Paid" : col_diff(col9, col3)
    })

    col12 = pd.DataFrame({
        "Actual IBNR" : col_diff(col_mult(ult, cdf), col3)
    })

    col13 = pd.DataFrame({
        "Difference from Actual IBNR - Reported" : col_diff(col12, col10)
    })

    col14 = pd.DataFrame({
        "Difference from Actual IBNR - Paid" : col_diff(col12, col11)
    })


    cols = [col1, col2, col3, col4, col5, col6, col7, col8, col9, col10, col11, col12, col13, col14]

    results = pd.concat(cols, axis=1)

    

    display(
        results
    )



### Exhibit III, Sheets 10 and 11: USPP Auto Development of Unpaid Claim Estimate

We then generate Sheets 10 and 11 which illustrate the differences between the actual IBNR and the IBNR obtained through the development method for each scenario (pp. 123-124).

In [ ]:
scenarios = ["Steady State","Increasing Claim","Increasing Case","Increasing Claim Case"]

for scenario in scenarios:
    Ex10Sht10(scenario, triangles)

,Accident Year,Age of Accident Year at 12/31/08,Claims at 12/31/2008 - Reported,Claims at 12/31/2008 - Paid,Case Outstanding,CDF to Ult. - Reported,CDF to Ult. - Paid,Projected Ult. Claims Using Dev. Method - Reported,Projected Ult. Claims Using Dev. Method - Paid,Estimated IBNR Using Dev. Method - Reported,Estimated IBNR Using Dev. Method - Paid,Actual IBNR,Difference from Actual IBNR - Reported,Difference from Actual IBNR - Paid
0,1999,120,"700,000","700,000",0,1,1,"700,000","700,000",0,0,0,0,0
1,2000,108,"735,000","735,000",0,1,1,"735,000","735,000",0,0,0,0,0
2,2001,96,"771,750","764,033","7,717",1,1,"771,750","771,751",0,1,0,0,-1
3,2002,84,"810,338","802,234","8,104",1,1,"810,338","810,337",0,-1,0,0,1
4,2003,72,"842,346","833,837","8,509",1,1,"850,855","850,854","8,509","8,508","8,509",0,0
5,2004,60,"884,463","857,661","26,802",1,1,"893,397","893,397","8,934","8,934","8,934",0,0
6,2005,48,"919,306","863,022","56,284",1,1,"938,068","938,067","18,762","18,761","18,762",0,0
7,2006,36,"935,722","827,375","108,347",1,1,"984,970","984,970","49,248","49,248","49,248",0,0
8,2007,24,"930,797","734,295","196,502",1,1,"1,034,219","1,034,218","103,422","103,421","103,422",0,1
9,2008,12,"836,166","456,090","380,076",1,2,"1,085,930","1,085,928","249,764","249,762","249,764",0,2


,Accident Year,Age of Accident Year at 12/31/08,Claims at 12/31/2008 - Reported,Claims at 12/31/2008 - Paid,Case Outstanding,CDF to Ult. - Reported,CDF to Ult. - Paid,Projected Ult. Claims Using Dev. Method - Reported,Projected Ult. Claims Using Dev. Method - Paid,Estimated IBNR Using Dev. Method - Reported,Estimated IBNR Using Dev. Method - Paid,Actual IBNR,Difference from Actual IBNR - Reported,Difference from Actual IBNR - Paid
0,1999,120,"700,000","700,000",0,1,1,"700,000","700,000",0,0,0,0,0
1,2000,108,"735,000","735,000",0,1,1,"735,000","735,000",0,0,0,0,0
2,2001,96,"771,750","764,033","7,717",1,1,"771,750","771,751",0,1,0,0,-1
3,2002,84,"810,338","802,234","8,104",1,1,"810,338","810,337",0,-1,0,0,1
4,2003,72,"842,346","833,837","8,509",1,1,"850,855","850,854","8,509","8,508","8,509",0,0
5,2004,60,"1,010,815","980,184","30,631",1,1,"1,021,025","1,021,025","10,210","10,210","10,210",0,0
6,2005,48,"1,116,300","1,047,955","68,345",1,1,"1,139,082","1,139,081","22,782","22,781","22,782",0,0
7,2006,36,"1,203,071","1,063,768","139,303",1,1,"1,266,391","1,266,390","63,320","63,319","63,320",0,0
8,2007,24,"1,263,224","996,544","266,680",1,1,"1,403,582","1,403,583","140,358","140,359","140,358",0,-0
9,2008,12,"1,194,523","651,558","542,965",1,2,"1,551,328","1,551,328","356,805","356,805","356,805",0,0


,Accident Year,Age of Accident Year at 12/31/08,Claims at 12/31/2008 - Reported,Claims at 12/31/2008 - Paid,Case Outstanding,CDF to Ult. - Reported,CDF to Ult. - Paid,Projected Ult. Claims Using Dev. Method - Reported,Projected Ult. Claims Using Dev. Method - Paid,Estimated IBNR Using Dev. Method - Reported,Estimated IBNR Using Dev. Method - Paid,Actual IBNR,Difference from Actual IBNR - Reported,Difference from Actual IBNR - Paid
0,1999,120,"700,000","700,000",0,1,1,"700,000","700,000",0,0,0,0,0
1,2000,108,"735,000","735,000",0,1,1,"735,000","735,000",0,0,0,0,0
2,2001,96,"771,750","764,033","7,717",1,1,"771,750","771,751",0,1,0,0,-1
3,2002,84,"810,338","802,234","8,104",1,1,"810,338","810,337",0,-1,0,0,1
4,2003,72,"842,346","833,837","8,509",1,1,"850,855","850,854","8,509","8,508","8,509",0,0
5,2004,60,"884,463","857,661","26,802",1,1,"893,397","893,397","8,934","8,934","8,934",0,0
6,2005,48,"933,377","863,022","70,355",1,1,"951,657","938,067","18,280","4,690","4,691","-13,589",0
7,2006,36,"962,808","827,375","135,433",1,1,"1,015,301","984,970","52,493","22,162","22,162","-30,331",0
8,2007,24,"979,922","734,295","245,627",1,1,"1,096,235","1,034,218","116,313","54,296","54,297","-62,016",1
9,2008,12,"931,185","456,090","475,095",1,2,"1,227,589","1,085,928","296,404","154,743","154,745","-141,659",2


,Accident Year,Age of Accident Year at 12/31/08,Claims at 12/31/2008 - Reported,Claims at 12/31/2008 - Paid,Case Outstanding,CDF to Ult. - Reported,CDF to Ult. - Paid,Projected Ult. Claims Using Dev. Method - Reported,Projected Ult. Claims Using Dev. Method - Paid,Estimated IBNR Using Dev. Method - Reported,Estimated IBNR Using Dev. Method - Paid,Actual IBNR,Difference from Actual IBNR - Reported,Difference from Actual IBNR - Paid
0,1999,120,"700,000","700,000",0,1,1,"700,000","700,000",0,0,0,0,0
1,2000,108,"735,000","735,000",0,1,1,"735,000","735,000",0,0,0,0,0
2,2001,96,"771,750","764,033","7,717",1,1,"771,750","771,751",0,1,0,0,-1
3,2002,84,"810,338","802,234","8,104",1,1,"810,338","810,337",0,-1,0,0,1
4,2003,72,"842,346","833,837","8,509",1,1,"850,855","850,854","8,509","8,508","8,509",0,0
5,2004,60,"1,010,815","980,184","30,631",1,1,"1,021,025","1,021,025","10,210","10,210","10,210",0,0
6,2005,48,"1,133,386","1,047,955","85,431",1,1,"1,155,482","1,139,081","22,096","5,695","5,696","-16,400",0
7,2006,36,"1,237,897","1,063,768","174,129",1,1,"1,305,639","1,266,390","67,742","28,493","28,494","-39,249",0
8,2007,24,"1,329,895","996,544","333,351",1,1,"1,488,875","1,403,583","158,980","73,688","73,687","-85,293",-0
9,2008,12,"1,330,264","651,558","678,706",1,2,"1,756,504","1,551,328","426,240","221,064","221,064","-205,176",0


## Exhibit IV — Impact of Changing Product Mix Example

In order to illustrate the assumptions and corresponding limitations of the Development method, Exhibit III applies the method to the **U.S. Private Passenger Industry Auto** (USPP Auto) example under four different scenarios:

1. stable claim ratios and no change in case outstanding strength (steady state);
1. increasing claim ratios but no change in case outstanding strength;
1. stable claim ratios but increasing case outstanding strength; and
1. increasing claim ratios and increasing case outstanding strength.

The key assumptions of this study are as follows:

- the earned premium for the first year (1999) is assumed to be $1 million with a 5% annual premium trend;
- (discuss assumed claim ratios)

Since the USPP Auto dataset is contained in the `chainladder` package, we simply load it into a triangle as follows:

In [65]:
tri_steady = cl.load_sample("friedland_us_auto_steady_state")
tri_mix = cl.load_sample("friedland_us_auto_chg_prod_mix")

tri_steady_rep = tri_steady["Reported Claims"]
tri_steady_paid = tri_steady["Paid Claims"]

tri_mix_rep = tri_mix["Reported Claims"]
tri_mix_paid = tri_mix["Paid Claims"]

cl.Development(n_periods = -1, average='volume', ).fit_transform(tri_steady_rep)


,12,24,36,48,60,72,84,96,108,120
1999,"1,011,000","1,254,000","1,377,000","1,454,000","1,477,000","1,493,000","1,500,000","1,500,000","1,500,000","1,500,000"
2000,"1,061,550","1,316,700","1,445,850","1,526,700","1,550,850","1,567,650","1,575,000","1,575,000","1,575,000",
2001,"1,114,628","1,382,535","1,518,143","1,603,035","1,628,393","1,646,033","1,653,750","1,653,750",,
2002,"1,170,359","1,451,662","1,594,050","1,683,187","1,709,812","1,728,334","1,736,438",,,
2003,"1,228,877","1,524,245","1,673,752","1,767,346","1,795,303","1,814,751",,,,
2004,"1,290,321","1,600,457","1,757,440","1,855,713","1,885,068",,,,,
2005,"1,354,837","1,680,480","1,845,312","1,948,499",,,,,,
2006,"1,422,579","1,764,504","1,937,577",,,,,,,
2007,"1,493,707",,,,,,,,,
2008,"3,421,122",,,,,,,,,
